// @metamask/snaps-sdk/dist/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assert = exports.nonEmptyRecord = exports.selectiveUnion = exports.typedUnion = exports.enumValue = exports.union = exports.literal = exports.SNAP_ERROR_MESSAGE = exports.SNAP_ERROR_CODE = exports.getErrorStack = exports.getErrorMessage = exports.getErrorData = void 0;
var internals_1 = require("./internals/index.cjs");
Object.defineProperty(exports, "getErrorData", { enumerable: true, get: function () { return internals_1.getErrorData; } });
Object.defineProperty(exports, "getErrorMessage", { enumerable: true, get: function () { return internals_1.getErrorMessage; } });
Object.defineProperty(exports, "getErrorStack", { enumerable: true, get: function () { return internals_1.getErrorStack; } });
Object.defineProperty(exports, "SNAP_ERROR_CODE", { enumerable: true, get: function () { return internals_1.SNAP_ERROR_CODE; } });
Object.defineProperty(exports, "SNAP_ERROR_MESSAGE", { enumerable: true, get: function () { return internals_1.SNAP_ERROR_MESSAGE; } });
Object.defineProperty(exports, "literal", { enumerable: true, get: function () { return internals_1.literal; } });
Object.defineProperty(exports, "union", { enumerable: true, get: function () { return internals_1.union; } });
Object.defineProperty(exports, "enumValue", { enumerable: true, get: function () { return internals_1.enumValue; } });
Object.defineProperty(exports, "typedUnion", { enumerable: true, get: function () { return internals_1.typedUnion; } });
Object.defineProperty(exports, "selectiveUnion", { enumerable: true, get: function () { return internals_1.selectiveUnion; } });
Object.defineProperty(exports, "nonEmptyRecord", { enumerable: true, get: function () { return internals_1.nonEmptyRecord; } });
var utils_1 = require("@metamask/utils");
Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return utils_1.assert; } });
__exportStar(require("./errors.cjs"), exports);
__exportStar(require("./error-wrappers.cjs"), exports);
__exportStar(require("./images.cjs"), exports);
__exportStar(require("./types/index.cjs"), exports);
__exportStar(require("./ui/index.cjs"), exports);
