// @metamask/snaps-sdk/dist/ui/nodes.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiteralStruct = exports.NodeStruct = exports.NodeType = void 0;
const superstruct_1 = require("@metamask/superstruct");
/**
 * The supported node types. This is based on SIP-7.
 *
 * @see https://metamask.github.io/SIPs/SIPS/sip-7
 */
var NodeType;
(function (NodeType) {
    NodeType["Copyable"] = "copyable";
    NodeType["Divider"] = "divider";
    NodeType["Heading"] = "heading";
    NodeType["Panel"] = "panel";
    NodeType["Spinner"] = "spinner";
    NodeType["Text"] = "text";
    NodeType["Image"] = "image";
    NodeType["Row"] = "row";
    NodeType["Address"] = "address";
    NodeType["Button"] = "button";
    NodeType["Input"] = "input";
    NodeType["Form"] = "form";
})(NodeType || (exports.NodeType = NodeType = {}));
/**
 * @internal
 */
exports.NodeStruct = (0, superstruct_1.object)({
    type: (0, superstruct_1.string)(),
});
/**
 * @internal
 */
exports.LiteralStruct = (0, superstruct_1.assign)(exports.NodeStruct, (0, superstruct_1.object)({
    value: (0, superstruct_1.unknown)(),
}));
