// @metamask/snaps-sdk/dist/jsx/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldChildUnionStruct = exports.FormChildStruct = exports.BoxChildStruct = exports.assertJSXElement = exports.isJSXElementUnsafe = exports.isJSXElement = exports.RootJSXElementStruct = exports.JSXElementStruct = void 0;
__exportStar(require("./component.cjs"), exports);
__exportStar(require("./components/index.cjs"), exports);
__exportStar(require("./jsx-runtime.cjs"), exports);
__exportStar(require("./jsx-dev-runtime.cjs"), exports);
var validation_1 = require("./validation.cjs");
Object.defineProperty(exports, "JSXElementStruct", { enumerable: true, get: function () { return validation_1.JSXElementStruct; } });
Object.defineProperty(exports, "RootJSXElementStruct", { enumerable: true, get: function () { return validation_1.RootJSXElementStruct; } });
Object.defineProperty(exports, "isJSXElement", { enumerable: true, get: function () { return validation_1.isJSXElement; } });
Object.defineProperty(exports, "isJSXElementUnsafe", { enumerable: true, get: function () { return validation_1.isJSXElementUnsafe; } });
Object.defineProperty(exports, "assertJSXElement", { enumerable: true, get: function () { return validation_1.assertJSXElement; } });
Object.defineProperty(exports, "BoxChildStruct", { enumerable: true, get: function () { return validation_1.BoxChildStruct; } });
Object.defineProperty(exports, "FormChildStruct", { enumerable: true, get: function () { return validation_1.FormChildStruct; } });
Object.defineProperty(exports, "FieldChildUnionStruct", { enumerable: true, get: function () { return validation_1.FieldChildUnionStruct; } });
