// @metamask/key-tree/dist/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mnemonicToSeed = exports.createBip39KeyFromSeed = exports.mnemonicPhraseToBytes = exports.isValidBIP32PathSegment = exports.getBIP44CoinTypeToAddressPathTuple = exports.getBIP44AddressKeyDeriver = exports.deriveBIP44AddressKey = exports.BIP_44_COIN_TYPE_DEPTH = exports.BIP44CoinTypeNode = exports.ed25519Bip32 = exports.ed25519 = exports.secp256k1 = exports.SLIP10Node = exports.BIP44Node = void 0;
var BIP44Node_1 = require("./BIP44Node.cjs");
Object.defineProperty(exports, "BIP44Node", { enumerable: true, get: function () { return BIP44Node_1.BIP44Node; } });
var SLIP10Node_1 = require("./SLIP10Node.cjs");
Object.defineProperty(exports, "SLIP10Node", { enumerable: true, get: function () { return SLIP10Node_1.SLIP10Node; } });
var curves_1 = require("./curves/index.cjs");
Object.defineProperty(exports, "secp256k1", { enumerable: true, get: function () { return curves_1.secp256k1; } });
Object.defineProperty(exports, "ed25519", { enumerable: true, get: function () { return curves_1.ed25519; } });
Object.defineProperty(exports, "ed25519Bip32", { enumerable: true, get: function () { return curves_1.ed25519Bip32; } });
var BIP44CoinTypeNode_1 = require("./BIP44CoinTypeNode.cjs");
Object.defineProperty(exports, "BIP44CoinTypeNode", { enumerable: true, get: function () { return BIP44CoinTypeNode_1.BIP44CoinTypeNode; } });
Object.defineProperty(exports, "BIP_44_COIN_TYPE_DEPTH", { enumerable: true, get: function () { return BIP44CoinTypeNode_1.BIP_44_COIN_TYPE_DEPTH; } });
Object.defineProperty(exports, "deriveBIP44AddressKey", { enumerable: true, get: function () { return BIP44CoinTypeNode_1.deriveBIP44AddressKey; } });
Object.defineProperty(exports, "getBIP44AddressKeyDeriver", { enumerable: true, get: function () { return BIP44CoinTypeNode_1.getBIP44AddressKeyDeriver; } });
__exportStar(require("./constants.cjs"), exports);
var utils_1 = require("./utils.cjs");
Object.defineProperty(exports, "getBIP44CoinTypeToAddressPathTuple", { enumerable: true, get: function () { return utils_1.getBIP44CoinTypeToAddressPathTuple; } });
Object.defineProperty(exports, "isValidBIP32PathSegment", { enumerable: true, get: function () { return utils_1.isValidBIP32PathSegment; } });
Object.defineProperty(exports, "mnemonicPhraseToBytes", { enumerable: true, get: function () { return utils_1.mnemonicPhraseToBytes; } });
var derivers_1 = require("./derivers/index.cjs");
Object.defineProperty(exports, "createBip39KeyFromSeed", { enumerable: true, get: function () { return derivers_1.createBip39KeyFromSeed; } });
Object.defineProperty(exports, "mnemonicToSeed", { enumerable: true, get: function () { return derivers_1.mnemonicToSeed; } });
